### PR TITLE
feat(devtools): Initial setup for usage telemetry

### PR DESCRIPTION
**Pull request by alexvy86**
## Description

Initial setup for enabling usage telemetry in the Fluid Devtools.

- New prop to pass an "external" logger to the top-level React component that renders the Devtools experience.
- New React context to provide an "internal" logger within the Devtools extension.
- Some convenience exports so consumers of `devtools-view` don't have to also import `common-definitions' to have access to the logger-related types.
- All telemetry events generated internally are automatically written to the console (only visible if the console is set to Verbose) besides being passed to any loggers provided "from outside". This makes them easily testable both in the extension and in the devtools-example where we render Devtools inline with the app. In the former case they show up in the extension's console (right-click somewhere in the rendered extension, click `Inspect`, switch to the Console tab) and in the latter they show up directly in the tab's console.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

[AB#4714](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4714)

### CODE REVIEW

**Comment by alexvy86** [comment]
@WayneFerrao particularly interested in your thoughts, to make sure this is clear and aligns with what we discussed recently.

With this PR in place, you could update `InitializeView.ts` in the extension package with something like this (plus the necessary import for `alertTelemetryLogger` included in this PR) to get telemetry events shown with `alert()` pop-ups in the browser:

image

I couldn't figure out a good and quick way to print them to console, but maybe someone will figure out how (without having to do message passing to the main tab so we print them from there).

File: packages/tools/devtools/devtools-view/src/LoggerContext.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Licensed under the MIT License.
+ */
+
+import React from "react";
+import { ITelemetryLoggerExt } from "@fluidframework/telemetry-utils";
+
+/**
+ * Context that provides a logger for Devtools to generate usage telemetry internally.
+ *
+ * @remarks
```
Comment (seanimam) [inline packages/tools/devtools/devtools-view/src/LoggerContext.ts:12]: Have not seen use of the @remarks jsdoc keyword before. I like this.

File: packages/tools/devtools/devtools-browser-extension/src/devtools/Logging.ts [hunk 2]
```diff
@@ -23,3 +24,18 @@ export const devtoolsScriptMessageLoggingOptions: MessageLoggingOptions = {
...
 	return `${devtoolsScriptLoggingContext}: ${text}`;
 }
+
+/**
+ * Logger that can be passed to the extension via a {@link @fluid-experimental/devtools-view#LoggerContext} to display
+ * usage telemetry events from the extension with alert popups in the browser.
+ *
+ * @remarks
+ * Only intended for testing the extension during development. Since the extension runs in a separate context than the
+ * current browser tab, console.log() does not display the messages anywhere we can use them for troubleshooting.
```
Comment (WayneFerrao) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/Logging.ts:34]: Might be a silly question but, can't we also right-click in the extension to inspect console and get a new console window for our extension in which console logs should appear?

**Comment by WayneFerrao** [comment]
@alexvy86 With respect to your question, I don't think alerts are needed. We can still use regular console logs if I've understood it correctly. For example, the console log in `InitializeView.ts` in devtools-browser-extension can be seen by inspecting the devtools extension.

**Comment by alexvy86** [comment]
> @alexvy86 With respect to your question, I don't think alerts are needed. We can still use regular console logs if I've understood it correctly. For example, the console log in `InitializeView.ts` in devtools-browser-extension can be seen by inspecting the devtools extension.

_That's_ the bit I was missing, thank you 😄. I hadn't tried to do Inspect on the extension itself. Updating the PR to use console.log there.

File: packages/tools/devtools/devtools-browser-extension/src/devtools/Logging.ts (hunk 2 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-browser-extension/src/devtools/Logging.ts]: Yes we can! TIL how to look at that console.

**Comment by alexvy86** [comment]
@WayneFerrao @seanimam I think this is done, ready for last review.

**Review by WayneFerrao — APPROVED** [review]
Looks good!

File: packages/tools/devtools/devtools-view/src/LoggerContext.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Gets the {@link @fluidframework/telemetry-utils#ITelemetryLoggerExt} provided through an {@link LoggerContext}.
+ *
+ * @returns
+ * The logger from the context, or undefined is no logger was provided.
+ */
+export function useLogger(): ITelemetryLoggerExt | undefined {
+	return React.useContext(LoggerContext);
+}
```
[the remark this answers is not in the document]
Comment (alexvy86) [reply packages/tools/devtools/devtools-view/src/LoggerContext.ts]: In this case we wouldn't be updating this particular context ever, so in practical terms it kind of acts like a singleton within devtools-view. I still lean towards this vs an actual singleton for testability; being able to pass in a mock logger through the context when necessary. We probably can figure out a way to do something similar with a singleton that can be configured "from outside", by the consumer of devtools-view, but I don't see a huge advantage to doing it that way.

**Pull request merged**